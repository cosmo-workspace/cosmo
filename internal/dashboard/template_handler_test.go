package dashboard

import (
	"context"
	"errors"
	"net/http"

	. "github.com/cosmo-workspace/cosmo/pkg/snap"
	. "github.com/onsi/ginkgo/v2"
	. "github.com/onsi/gomega"

	"google.golang.org/protobuf/types/known/emptypb"

	cosmov1alpha1 "github.com/cosmo-workspace/cosmo/api/v1alpha1"
	"github.com/cosmo-workspace/cosmo/proto/gen/dashboard/v1alpha1/dashboardv1alpha1connect"
)

var _ = Describe("Dashboard server [Template]", func() {

	var (
		userSession  string
		adminSession string
		client       dashboardv1alpha1connect.TemplateServiceClient
	)

	BeforeEach(func() {
		userSession = test_CreateLoginUserSession("normal-user", "お名前", nil, "password")
		adminSession = test_CreateLoginUserSession("admin-user", "アドミン", []cosmov1alpha1.UserRole{cosmov1alpha1.PrivilegedRole}, "password")
		client = dashboardv1alpha1connect.NewTemplateServiceClient(http.DefaultClient, "http://localhost:8888")
	})

	AfterEach(func() {
		clientMock.Clear()
		testUtil.DeleteCosmoUserAll()
		testUtil.DeleteTemplateAll()
	})

	Describe("[GetWorkspaceTemplates]", func() {

		run_test := func(loginUser string, testCase string) {
			if testCase == "not empty" {
				testUtil.CreateTemplate(cosmov1alpha1.TemplateLabelEnumTypeWorkspace, "template1")
				testUtil.CreateTemplate(cosmov1alpha1.TemplateLabelEnumTypeWorkspace, "template2")
			}
			session := userSession
			if loginUser == "admin-user" {
				session = adminSession
			}
			By("---------------test start----------------")
			ctx := context.Background()
			res, err := client.GetWorkspaceTemplates(ctx, NewRequestWithSession(&emptypb.Empty{}, session))
			if err == nil {
				Ω(res.Msg).To(MatchSnapShot())
			} else {
				Ω(err.Error()).To(MatchSnapShot())
				Expect(res).Should(BeNil())
			}
			By("---------------test end---------------")
		}

		DescribeTable("✅ success in normal context:",
			run_test,
			Entry(nil, "admin-user", "empty"),
			Entry(nil, "admin-user", "not empty"),
			Entry(nil, "normal-user", "empty"),
			Entry(nil, "normal-user", "not empty"),
		)

		DescribeTable("❌ fail with an unexpected error at list:",
			func(user string, testCase string) {
				clientMock.SetListError((*Server).GetWorkspaceTemplates, errors.New("template list error"))
				run_test(user, testCase)
			},
			Entry(nil, "admin-user", "not empty"),
		)
	})

	Describe("[GetUserAddonTemplates]", func() {

		run_test := func(loginUser string, testCase string) {
			if testCase == "not empty" {
				testUtil.CreateTemplate(cosmov1alpha1.TemplateLabelEnumTypeUserAddon, "useraddon1")
				testUtil.CreateTemplate(cosmov1alpha1.TemplateLabelEnumTypeUserAddon, "useraddon2")
			}
			session := userSession
			if loginUser == "admin-user" {
				session = adminSession
			}
			By("---------------test start----------------")
			ctx := context.Background()
			res, err := client.GetUserAddonTemplates(ctx, NewRequestWithSession(&emptypb.Empty{}, session))
			if err == nil {
				Ω(res.Msg).To(MatchSnapShot())
			} else {
				Ω(err.Error()).To(MatchSnapShot())
				Expect(res).Should(BeNil())
			}
			By("---------------test end---------------")
		}

		DescribeTable("✅ success in normal context:",
			run_test,
			Entry(nil, "admin-user", "empty"),
			Entry(nil, "admin-user", "not empty"),
			Entry(nil, "normal-user", "empty"),
			Entry(nil, "normal-user", "not empty"),
		)

		DescribeTable("❌ fail with an unexpected error at list:",
			func(user string, testCase string) {
				clientMock.SetListError((*Server).GetUserAddonTemplates, errors.New("template list error"))
				run_test(user, testCase)
			},
			Entry(nil, "admin-user", "not empty"),
		)
	})
})
