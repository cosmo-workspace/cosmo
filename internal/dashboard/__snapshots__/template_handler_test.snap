['Dashboard server [Template] [GetUserAddonTemplates] ✅ success in normal context: Entry: admin-user, empty 1']
SnapShot = """
{
  \"message\": \"No items found\"
}
"""

['Dashboard server [Template] [GetUserAddonTemplates] ✅ success in normal context: Entry: admin-user, not empty 1']
SnapShot = """
{
  \"items\": [
    {
      \"name\": \"useraddon1\",
      \"required_vars\": [
        {
          \"var_name\": \"{{HOGE}}\",
          \"default_value\": \"HOGEhoge\"
        },
        {
          \"var_name\": \"{{FUGA}}\",
          \"default_value\": \"FUGAfuga\"
        }
      ],
      \"is_default_user_addon\": true
    },
    {
      \"name\": \"useraddon2\",
      \"required_vars\": [
        {
          \"var_name\": \"{{HOGE}}\",
          \"default_value\": \"HOGEhoge\"
        },
        {
          \"var_name\": \"{{FUGA}}\",
          \"default_value\": \"FUGAfuga\"
        }
      ],
      \"is_default_user_addon\": true
    }
  ]
}
"""

['Dashboard server [Template] [GetUserAddonTemplates] ✅ success in normal context: Entry: normal-user, empty 1']
SnapShot = """
{
  \"message\": \"No items found\"
}
"""

['Dashboard server [Template] [GetUserAddonTemplates] ✅ success in normal context: Entry: normal-user, not empty 1']
SnapShot = """
{
  \"items\": [
    {
      \"name\": \"useraddon1\",
      \"required_vars\": [
        {
          \"var_name\": \"{{HOGE}}\",
          \"default_value\": \"HOGEhoge\"
        },
        {
          \"var_name\": \"{{FUGA}}\",
          \"default_value\": \"FUGAfuga\"
        }
      ],
      \"is_default_user_addon\": true
    },
    {
      \"name\": \"useraddon2\",
      \"required_vars\": [
        {
          \"var_name\": \"{{HOGE}}\",
          \"default_value\": \"HOGEhoge\"
        },
        {
          \"var_name\": \"{{FUGA}}\",
          \"default_value\": \"FUGAfuga\"
        }
      ],
      \"is_default_user_addon\": true
    }
  ]
}
"""

['Dashboard server [Template] [GetUserAddonTemplates] ❌ fail with an unexpected error at list: Entry: admin-user, not empty 1']
SnapShot = 'internal: failed to list UserAddon Templates'

['Dashboard server [Template] [GetWorkspaceTemplates] ✅ success in normal context: Entry: admin-user, empty 1']
SnapShot = """
{
  \"message\": \"No items found\"
}
"""

['Dashboard server [Template] [GetWorkspaceTemplates] ✅ success in normal context: Entry: admin-user, not empty 1']
SnapShot = """
{
  \"items\": [
    {
      \"name\": \"template1\",
      \"required_vars\": [
        {
          \"var_name\": \"{{HOGE}}\",
          \"default_value\": \"HOGEhoge\"
        },
        {
          \"var_name\": \"{{FUGA}}\",
          \"default_value\": \"FUGAfuga\"
        }
      ]
    },
    {
      \"name\": \"template2\",
      \"required_vars\": [
        {
          \"var_name\": \"{{HOGE}}\",
          \"default_value\": \"HOGEhoge\"
        },
        {
          \"var_name\": \"{{FUGA}}\",
          \"default_value\": \"FUGAfuga\"
        }
      ]
    }
  ]
}
"""

['Dashboard server [Template] [GetWorkspaceTemplates] ✅ success in normal context: Entry: normal-user, empty 1']
SnapShot = """
{
  \"message\": \"No items found\"
}
"""

['Dashboard server [Template] [GetWorkspaceTemplates] ✅ success in normal context: Entry: normal-user, not empty 1']
SnapShot = """
{
  \"items\": [
    {
      \"name\": \"template1\",
      \"required_vars\": [
        {
          \"var_name\": \"{{HOGE}}\",
          \"default_value\": \"HOGEhoge\"
        },
        {
          \"var_name\": \"{{FUGA}}\",
          \"default_value\": \"FUGAfuga\"
        }
      ]
    },
    {
      \"name\": \"template2\",
      \"required_vars\": [
        {
          \"var_name\": \"{{HOGE}}\",
          \"default_value\": \"HOGEhoge\"
        },
        {
          \"var_name\": \"{{FUGA}}\",
          \"default_value\": \"FUGAfuga\"
        }
      ]
    }
  ]
}
"""

['Dashboard server [Template] [GetWorkspaceTemplates] ❌ fail with an unexpected error at list: Entry: admin-user, not empty 1']
SnapShot = 'internal: failed to list WorkspaceTemplates'
